pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -571,6 +575,17 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
             printSendCommandDebug(send, headersAndPayload);
         }
 
+        // avoid processing non-persist message if reached max concurrent-message limit
+        if (producer.isNonPersistentTopic() && !nonPersistentMessageSemaphore.tryAcquire()) {
+            final long producerId = send.getProducerId();
+            final long sequenceId = send.getSequenceId();
+            service.getTopicOrderedExecutor().submitOrdered(producer.getTopic(), SafeRun.safeRun(() -> {
```

Comment:
Uhm, but wait for dispatching message to consumers before sending ack to producer? 

I think the best option here is to send ack immediately to producer. We got the message in broker memory, that's the level of guarantee. 

Later if the consumers are not ready, the message will be dropped.